www/qserv/js/ToolsSql.js

```diff
@@ -0,0 +1,219 @@
+    <div class="col-3">
+      <input type="text" class="form-control" placeholder="Worker" id="worker">
+    </div>
+    <div class="col-3">
+      <input type="text" class="form-control" placeholder="User" id="user">
+    </div>
+    <div class="col-3">
+      <input type="password" class="form-control" placeholder="Password" id="password">
+    </div>
+    <div class="col-3">
+      <button class="btn btn-primary   btn-sm" id="execute">Execute</button>
+      <button class="btn btn-secondary btn-sm" id="reset"  >Reset form</button>
+    </div>
+  </div>
+  <div class="row">
+    <div class="col">
+      <textarea rows="3" class="form-control" placeholder="Query" id="query"></textarea>
+    </div>
+  </div>
+  <div class="row">
+    <div class="col">
+      <table class="table table-sm table-hover table-bordered" id="fwk-tools-sql-resultset">
+        <caption></caption>
+        <thead class="thead-light">
+        </thead>
+        <tbody></tbody>
+      </table>
+    </div>
+  </div>
+</div>`;
+            this.fwk_app_container.html(html);
+            this.fwk_app_container.find('button#execute').click(() => {
+                this._load();
+            });
+            this.fwk_app_container.find('button#reset').click(() => {
+                this._worker().val('');
+                this._user().val('');
+                this._password().val('');
+                this._query().val('');
+            });
+        }
+
+        // Inputs
+
+        _worker() {
+            if (this._worker_obj === undefined) {
+                this._worker_obj = this.fwk_app_container.find('input#worker');
+            }
+            return this._worker_obj;
+        }
+        _user() {
+            if (this._user_obj === undefined) {
+                this._user_obj = this.fwk_app_container.find('input#user');
+            }
+            return this._user_obj;
+        }
+        _password() {
+            if (this._password_obj === undefined) {
+                this._password_obj = this.fwk_app_container.find('input#password');
+            }
+            return this._password_obj;
+        }
+        _query() {
+            if (this._query_obj === undefined) {
+                this._query_obj = this.fwk_app_container.find('textarea#query');
+            }
+            return this._query_obj;
+        }
+
+        /**
+         * The table for displaying result sets of queries
+         * @returns JQuery table object
+         */
+        _table() {
+            if (this._table_obj === undefined) {
+                this._table_obj = this.fwk_app_container.find('table#fwk-tools-sql-resultset');
+            }
+            return this._table_obj;
+        }
+
+        /**
+         * Load data from a web servie then render it to the application's page.
+         */
+        _load() {
+            if (this._loading === undefined) {
+                this._loading = false;
+            }
+            if (this._loading) return;
+            this._loading = true;
```

Comment:
Isn't there a race condition here? Does JS have mutexes to protect it? Does this question even make sense? 😈

Reply:
JS is:
- single-threaded
- event-driven
Hence, no synchronization is needed. Each event handler executes within a safe environment. In fact, every code sequence  in JS is an "event". Events are serialized by a browser.